File: JavaScript/6kyu/SortTheOdd.js
function sortArray(array) {
  var odds = array.filter((x) => x % 2 != 0).sort((a, b) => b - a);

  return array.map((a) => (a % 2 ? odds.pop() : a));
}

function sortArray_kata_1(array) {
  var odds = array.filter((x) => x % 2 != 0).sort((a, b) => a - b);

  return array.map((a) => (a % 2 ? odds.shift() : a));
}

var input = [
  47, -18, 20, -10, 11, 12, 4, -16, -36, -13, -14, -50, -14, 34, 41, -38, -45,
];
var result = sortArray_kata_1(input);

console.log(result, "Result");
console.log(
  [-45, -18, 20, -10, -13, 12, 4, -16, -36, 11, -14, -50, -14, 34, 41, -38, 47],
  "Exepted"
);

File: Python/6kyu/SortTheOdd.py
import time
from itertools import cycle
# Number may repeat !!!


def sort_array(source_array):
    sorted_array = sorted([x for x in source_array if x % 2 != 0])
    arr = cycle(sorted_array)

    for i, x in enumerate(source_array):
        if x % 2 != 0:
            source_array[i] = next(arr)

    return source_array


def sort_array_kata_1(arr):
    odds = sorted((x for x in arr if x % 2 != 0), reverse=True)

    return [x if x % 2 == 0 else odds.pop() for x in arr]


def sort_array_kata_2(arr):
    odds = iter(sorted(v for v in arr if v % 2))
    return [next(odds) if i % 2 else i for i in arr]


input = [25, -33, -33, -5, 12, -5, -33, -7, -9, -
         10, -45, 43, -18, 36, 3, -28, 19, 16, 12, 3, 47, 24]


result = sort_array_kata_2(input)
print(result)
